app/frontend/src/features/profile/view/leaveReference/ReferenceForm.tsx

```diff
@@ -38,72 +39,85 @@ export const useReferenceStyles = makeStyles((theme) => ({
+  setReferenceValues: (values: ReferenceContextFormData) => void;
+}
+
+interface ReferenceRouteParams {
+  referenceType: string;
+  userId: string;
+  hostRequest?: string;
+  step?: string;
+}
+
+interface ReferenceFormProps {
+  user: User.AsObject;
+}
+
 export default function ReferenceForm({ user }: ReferenceFormProps) {
-  const { referenceType } = useParams<{
-    referenceType: string;
-  }>();
+  const [referenceData, setReferenceData] = useState<ReferenceContextFormData>({
+    text: "",
+    wasAppropriate: "",
+    rating: 0,
+  });
+
+  const setReferenceValues = (values: ReferenceContextFormData) => {
+    setReferenceData((prevData) => ({
+      ...prevData,
+      ...values,
+    }));
+  };
+
+  const friendMatch = useRouteMatch<ReferenceRouteParams>({
+    path: `${leaveReferenceBaseRoute}/:referenceType/:userId/:step?`,
+  });
+  const hostingMatch = useRouteMatch<ReferenceRouteParams>({
+    path: `${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequestId/:step?`,
+  });
+  const allMatch = useRouteMatch<ReferenceRouteParams>({
+    path: `${leaveReferenceBaseRoute}/:referenceType/:userId`,
+  });
+
+  const step =
+    allMatch?.params.referenceType === referenceTypeRoute[0]
+      ? friendMatch?.params.step
+      : hostingMatch?.params.step;
 
-  return (
-    <ReferenceDataProvider>
-      {referenceType === "friend" ? (
-        <Switch>
-          <Route
-            exact
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId`}
-          >
-            <Appropriate user={user} />
-          </Route>
-          <Route
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId/rating`}
-          >
-            <Rating user={user} />
-          </Route>
-          <Route
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId/reference/`}
-          >
-            <Text user={user} />
-          </Route>
-          <Route
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId/submit`}
-          >
-            <SubmitFriendReference user={user} />
-          </Route>
-        </Switch>
-      ) : (
-        <Switch>
-          <Route
-            exact
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequest`}
-          >
-            <Appropriate user={user} />
-          </Route>
-          <Route
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequest/rating`}
-          >
-            <Rating user={user} />
-          </Route>
-          <Route
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequest/reference`}
-          >
-            <Text user={user} />
-          </Route>
-          <Route
-            path={`${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequest/submit`}
-          >
-            <SubmitHostRequestReference user={user} />
-          </Route>
-        </Switch>
-      )}
-    </ReferenceDataProvider>
+  return step === undefined ? (
```

Comment:
This is great! I have actually never used `useRouteMatch` before and wasn't sure it would work 🤣

Reply:
Yeah, great call! It made it a lot cleaner